proposals/csharp-10.0/using-alias-improvements.md

```diff
@@ -0,0 +1,117 @@
+﻿# Using alias improvements
+
+## Summary
+
+Extend `using` aliases to allow generic type aliases and additional alias target types.
+
+## Generic type aliases
+
+Type aliases may include type parameters.
+```C#
+using MyList<T> = System.Collections.Generic.List<T>; // ok
+```
+
+Type parameters do not need to be referenced in the target.
+```C#
+using MyType1<T> = System.String; // ok
+using MyType2<T, U> = T;          // ok
+```
+
+Type parameter references in an alias target have the same restrictions as type parameter references in other generic contexts.
+```C#
+using MyType3<T, U> = T.U;  // error: type or namespace 'T' not found
+using MyType3<T, U> = T<U>; // error: type or namespace 'T<>' not found
+```
+
+The same name may be used for multiple aliases with distinct arity.
+```C#
+using MyDictionary = System.Collections.Generic.Dictionary<string, object>;
+using MyDictionary<T> = System.Collections.Generic.Dictionary<string, T>; // ok
+using MyDictionary<U> = System.Collections.Generic.Dictionary<string, U>; // error: 'MyDictionary<>' already defined
+```
+
+A partially bound type is not valid.
+The `using` aliases below are both valid, and the use of `MyList<>` is valid because the expansion `List<>` is an unbound type, but the use of `MyDictionary<>` invalid because `Dictionary<string,>` is a partially bound type.
+```C#
+using MyList<T> = System.Collections.Generic.List<T>;
+using MyDictionary<T> = System.Collections.Generic.Dictionary<string, T>;
+
+_ = typeof(MyList<>);       // ok: 'List<>'
+_ = typeof(MyDictionary<>); // error: 'Dictionary<string,>' is not valid
+```
+
+### Constraints
+If type parameter constraints cannot be declared explicitly, then constraints in the alias target are verified at the use site only.
+```C#
+using Option<T> = System.Nullable<T>; // ok
+static void F<T>(Option<T> t) { ... } // error: 'T' must be value type in 'Nullable<T>'
+```
+
+However, without explicit constraints, the compiler will treat `T?` as `Nullable<T>` in an alias target which means nullable reference types cannot be used with type parameters in aliases.
```

Comment:
Should we even permit `T?` at all? Or if we do, is this line even true? If we view these as macros, then `T?` will mean whatever `T?` would mean to the final expansion, not anything predefined like `Nullable<T>`. For example, `using MyMap<T> = Dictionary<T, T?>;` would cause different expansions for `int` and `string`, where one would be mapped to `int?` and one would be mapped to `string?`.